internal/chain/reward.go

```diff
@@ -1,157 +1,312 @@
+		bc.CurrentHeader().ShardID() == shard.BeaconChainShardID {
 
-	for idx, member := range parentCommittee.Slots {
-		switch signed, err := mask.IndexEnabled(idx); true {
-		case err != nil:
-			return ctxerror.New("cannot check for committer bit",
-				"committerIndex", idx,
-			).WithCause(err)
-		case signed:
-			accounts = append(accounts, member.EcdsaAddress)
-		default:
-			missing = append(missing, member)
+		// Take care of my own beacon chain committee, _ is missing, for slashing
+		members, payable, _, err := ballotResultBeaconchain(bc, header)
+		if err != nil {
+			return err
 		}
-	}
 
-	// do it quickly
-	w := sync.WaitGroup{}
-	for i := range missing {
-		w.Add(1)
-		go func(member int) {
-			defer w.Done()
-			// Slash if missing block was long enough
-			if slasher.ShouldSlash(missing[member].BlsPublicKey) {
-				// TODO Logic
+		votingPower := votepower.Compute(members)
+
+		for beaconMember := range payable {
+			voter := votingPower.Voters[payable[beaconMember].BlsPublicKey]
+			if !voter.IsHarmonyNode {
+				due := BlockRewardStakedCase.Mul(voter.EffectivePercent)
+				state.AddBalance(voter.EarningAccount, due.RoundInt())
 			}
-		}(i)
-	}
+		}
 
-	w.Wait()
+		// Handle rewards for shardchain
+		if cxLinks := header.CrossLinks(); len(cxLinks) != 0 {
+			crossLinks := types.CrossLinks{}
+			err := rlp.DecodeBytes(cxLinks, &crossLinks)
+			if err != nil {
+				fmt.Println("cross-link-error", err)
+				return err
+			}
+
+			w := sync.WaitGroup{}
+
+			type slotPayable struct {
+				effective numeric.Dec
+				payee     common.Address
+				nonce     int
+				oops      error
+			}
+
+			payable := make(chan slotPayable)
+
+			slotError := func(err error, receive chan slotPayable) {
+				s := slotPayable{}
+				s.oops = err
+				go func() {
+					receive <- s
+				}()
+			}
+
+			for i := range crossLinks {
+				w.Add(1)
+
+				go func(i int) {
+					defer w.Done()
+					cxLink := crossLinks[i]
+					subCommittee := shard.State{}
+					if err := rlp.DecodeBytes(
+						cxLink.ChainHeader.ShardState(), &subCommittee,
+					); err != nil {
+						slotError(err, payable)
+						return
+					}
+
+					subComm := subCommittee.FindCommitteeByID(cxLink.ShardID())
+					// _ are the missing signers, later for slashing
+					payableSigners, _, err := blockSigners(cxLink.Header(), subComm)
+					votingPower := votepower.Compute(subComm.Slots)
+
+					if err != nil {
+						slotError(err, payable)
+						return
+					}
+
+					for member := range payableSigners {
+						voter := votingPower.Voters[payableSigners[member].BlsPublicKey]
+						if !voter.IsHarmonyNode {
+							due := BlockRewardStakedCase.Mul(voter.EffectivePercent)
+							to := voter.EarningAccount
+							go func(signersDue numeric.Dec, addr common.Address, j int) {
+								payable <- slotPayable{
+									effective: signersDue,
+									payee:     addr,
+									nonce:     i * (i + j),
+									oops:      nil,
+								}
+							}(due, to, member)
```

Comment:
just give the nonce directly here. the formula i * (i + j) doesn't work. The sizes of payableSigners varies. No single formula does the work. Maybe just make all this single threaded?